napari/_qt/qt_resources/_icons.py

```diff
@@ -0,0 +1,317 @@
+    # handling of "shared" temporary files.
+    tf = NamedTemporaryFile(suffix='.py', delete=False)
+    try:
+        tf.close()
+        processResourceFile([qrc], tf.name, False)
+        out = Path(tf.name).read_bytes()
+    finally:
+        os.unlink(tf.name)
+    return out
+
+
+def _compile_qrc_pyside2(qrc) -> bytes:
+    """Compile qrc file using the PySide2 method.
+
+    PySide compiles qrc files using the rcc binary in the root directory, (same
+    path as PySide2.__init__)
+    """
+    from subprocess import CalledProcessError, run
+
+    import PySide2
+
+    pyside_root = Path(PySide2.__file__).parent
+
+    if os.name == 'nt':
+        look_for = ('rcc.exe', 'pyside2-rcc.exe')
+    else:
+        look_for = ('rcc', 'pyside2-rcc')
+
+    for bin in look_for:
+        if (pyside_root / bin).exists():
+            cmd = [str(pyside_root / bin)]
+            if 'pyside2' not in bin:
+                # the newer pure rcc version requires this for python
+                cmd.extend(['-g', 'python'])
+            break
+    else:
+        raise RuntimeError(f"PySide2 rcc binary not found in {pyside_root}")
+
+    try:
+        return run(cmd + [qrc], check=True, capture_output=True).stdout
+    except CalledProcessError as e:
+        raise RuntimeError(f"Failed to build PySide2 resources {e}")
+
+
+def compile_qrc(qrc) -> bytes:
+    """Compile a qrc file into a resources.py bytes"""
+    if qtpy.API_NAME == 'PyQt5':
+        return _compile_qrc_pyqt5(qrc).replace(b'PyQt5', b'qtpy')
+    elif qtpy.API_NAME == 'PySide2':
+        return _compile_qrc_pyside2(qrc).replace(b'PySide2', b'qtpy')
+    else:
+        raise RuntimeError(
+            f"Cannot compile QRC. Unexpected qtpy API name: {qtpy.API_NAME}"
+        )
+
+
+def compile_qt_svgs(
+    svg_paths: Iterable[Union[str, Path]],
+    colors: Iterable[Union[str, Tuple[str, str]]],
+    opacities: Iterable[float] = (1.0,),
+    theme_override: Optional[Dict[str, str]] = None,
+    prefix='',
+    save_path: Optional[str] = None,
+    register: bool = False,
+) -> str:
+    """[summary]
+
+    Parameters
+    ----------
+    svg_paths : Iterable[Union[str, Path]]
+        An iterable of paths to svg files
+    colors : Iterable[Union[str, Tuple[str, str]]]
+        An iterable of colors.  Every icon will be generated in every color. If
+        a `color` item is a string, it should be valid svg color style. Items
+        may also be a 2-tuple of strings, in which case the first item should
+        be an available theme name
+        (:func:`~napari.utils.theme.available_themes`), and the second item
+        should be a key in the theme (:func:`~napari.utils.theme.get_theme`),
+    opacities : Iterable[float], optional
+        An iterable of opacities to generate, by default (1.0,) Opacities less
+        than one can be accessed in qss with the opacity as a percentage
+        suffix, e.g.: `my_svg_50.svg` for opacity 0.5.
+    theme_override : dict, optional
```

Comment:
I see that we're using this internally, and I think i get what it is doing at the level of code and dicts, but I don't quite get why we need it. Why can't we just get the color right the first time :-)

Also is the first part of the dict the name of the icon without `.svg` at the final level of the path? I wonder then if we should be doing more with dicts like that instead of iterables for the other parameters too.

Maybe @jni has ideas here too? Note I don't mind this, just trying to think it through.

Reply:
> I don't quite get why we need it. Why can't we just get the color right the first time :-)

yeah, it's a good question.  The point is not _all_ icons necessarily want to be the "icon" color from the theme. Some (like the "warning.svg" icon) want to stand out, and use the "warning" color from the theme.  So, to get it "right" the first time, we either need to literally map every icon-name to ultimate color (and remove the "icon" color from the theme dict)... or say "everything svg should be the 'icon' color, unless otherwise specified", which is what the override is doing.

> I wonder then if we should be doing more with dicts like that instead of iterables for the other parameters too.

not sure I'm following what changes elsewhere you'd be referring to

Reply:
ah yes i get it now

> not sure I'm following what changes elsewhere you'd be referring to

yeah never mind, wasn't a good idea